### PR TITLE
TypeScript: Support a TS 3.9 breaking change for Optional Chaining and Non-Null Assertions

### LINKED ISSUE #8339: Parenthesized non-null assertion in an optional chain
**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAKAHgfgHQE8CUAhNgF4gA0IEADjAJbQDOyoAhgE7sQDuAChwmYpWAG26tczSgCN2rMAGs4MAMrV5dKAHNkMdgFc4lABYwAtiIDqxuvEbqwcFYNt0AbrdzJwjKSE2McOwwvHJaZqzIAGaigZQAVozoAEJyisoqrGZwADKacNGxRiCJ6CqaWiJwAIr6EPCFInEg6uyB7N4wuNRwjGDsdLQULQOwlnQAJjDGyAAcAAyU1FyBlnLU3su9QW4FlACOdfChNEIgrIwAtFBwcBN3w+xwh3RPoazhkUgxTcWBZnRdAY-hUqrV6gVvkVKDBWNJxlMZkgAEwwuR0EQVADCEDMEW8vQArMN9IEACpwoQ-ZpuQwASSg91gKn6gxgAEFGSoulVGoEAL78oA)

```sh
# Options (if any):
--parser=typescript
```

**Input:**

```jsx
(x?.y)!.z
```

**Output:**

```jsx
x?.y!.z;
```

**Expected behavior:**

```jsx
(x?.y)!.z
```

---

As of [TS 3.9](https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/#breaking-changes), `x?.y!.z` no longer means `(x?.y).z`.
Instead, the `!` is applied solely to the type of the property `y`, preserving the optionality of the chain (i.e. `x?.y!.z` is the same as `x?.y.z`, and `(x?.y)!.z` is the same as `(x?.y).z`).

This means that now automatically removing the parentheses in this case is wrong, and changes the meaning of the code.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fixes #8339

- https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/#breaking-changes
- https://babeljs.io/blog/2020/05/25/7.10.0#improved-optional-chaining-ergonomics-10961-https-githubcom-babel-babel-pull-10961-11248-https-githubcom-babel-babel-pull-11248 

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -626,6 +626,14 @@ function needsParens(path, options) {
 
     case "OptionalMemberExpression":
     case "OptionalCallExpression":
+      if (
+        node.type === "OptionalMemberExpression" &&
```
Comment (thorn0) [inline src/language-js/needs-parens.js:630]: `OptionalCallExpression` should be handled too.

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -626,6 +626,14 @@ function needsParens(path, options) {
...
+        parent.type === "TSNonNullExpression" &&
+        path.getParentNode(1).type === "MemberExpression"
```
Comment (thorn0) [inline src/language-js/needs-parens.js:632]: Member expressions have two child nodes, so it's not enough to check only `type` here.

File: src/language-js/needs-parens.js [hunk 3]
```diff
@@ -626,6 +626,14 @@ function needsParens(path, options) {
...
+      ) {
+        return true;
+      }
+
       if (
         (parent.type === "MemberExpression" && name === "object") ||
```
Comment (thorn0) [inline src/language-js/needs-parens.js:638]: Please add your check to the existing `if`.

File: src/language-js/needs-parens.js (hunk 2 above)
Comment (fisker) [reply src/language-js/needs-parens.js:632]: This?

```ts
(a?.b)![a?.b!]
```

Comment (thorn0) [reply src/language-js/needs-parens.js:632]: yep

**Pull request merged**